File: Army(java)/Magic/Spell.java
package Magic;

import Basic.Unit;
import Basic.*;
import State.*;
import EnumID.*;

public class Spell {
    protected int manaCost;
    protected int mana;
    protected int hitPoints;

    public Spell(int cost, int mana, int hp) {
        this.manaCost = cost;
        this.mana = mana;
        this.hitPoints = hp;
    }

    public void mAttack(Unit enemy, SpellCaster magician) throws UnitIsDeadException, MagicManaOffException {
        magician.getHealth().ensureIsAlive();
        magician.getMagicPower().ensureHaveIsMana();
        enemy.getHealth().ensureIsAlive();
        
        if ( magician.getMagicPower().getMana() < manaCost) {
            System.out.println("Not enough mana!");
            return;
        }
        
        if ( enemy.getWarriorID() == WarriorID.BERSERKER ) {
            System.out.println("Magic is powerless!");
            return;
        }
        
        if ( magician.getMagicianID() == MagicianID.SORCERER ) {
            enemy.getHealth().takeDamage(hitPoints);
        }
        
        if ( magician.getMagicianID() == MagicianID.HEALING ) {
            enemy.getHealth().takeDamage(hitPoints/2);
        }
        
        if ( magician.getMagicianID() == MagicianID.CHURCH_WORKER ) {
            if ( enemy.getWarriorID() == WarriorID.VAMPIRE || enemy.getWarriorID() == WarriorID.NECROMANCER ) {
                enemy.getHealth().takeDamage(hitPoints);
            } else {
                enemy.getHealth().takeDamage(hitPoints/2);
            }
        }
        
        magician.getMagicPower().takeMana(manaCost);
        
        enemy.notifyObserver();
    }

    public void healHealth(Unit other, SpellCaster magician) throws UnitIsDeadException, MagicManaOffException {
        magician.getHealth().ensureIsAlive();
        
        if ( magician.getMagicPower().getMana() < manaCost) {
            System.out.println("Not enough mana!");
            return;
        }
        
        if ( other.getWarriorID() == WarriorID.BERSERKER || other.getWarriorID() == WarriorID.DEMON ) {
            System.out.println("Magic is powerless!");
            return;
        }
        
        if ( magician.getMagicianID() == MagicianID.HEALING || magician.getMagicianID() == MagicianID.CHURCH_WORKER ) {
            other.getHealth().addHitPoints(hitPoints);
        } else {
            other.getHealth().addHitPoints(hitPoints/2);
        }
        
        magician.getMagicPower().takeMana(manaCost);
    }

    public void healMagic(SpellCaster other, SpellCaster magician) throws UnitIsDeadException, MagicManaOffException {
        magician.getHealth().ensureIsAlive();
        other.getHealth().ensureIsAlive();
        
        if ( magician.getMagicPower().getMana() < manaCost) {
            System.out.println("Not enough mana!");
            return;
        }
        
        if ( magician.getMagicianID() == MagicianID.HEALING || magician.getMagicianID() == MagicianID.CHURCH_WORKER ) {
            other.getMagicPower().addMana(mana);
        } else {
            other.getMagicPower().addMana(mana/2);
        }
        
        magician.getMagicPower().takeMana(manaCost);
    }

    public void backToHuman(Unit enemy, SpellCaster magician) throws UnitIsDeadException, MagicManaOffException {
        magician.getHealth().ensureIsAlive();
        magician.getMagicPower().ensureHaveIsMana();
        enemy.getHealth().ensureIsAlive();
        
        if ( magician.getMagicPower().getMana() < manaCost) {
            System.out.println("Not enough mana!");
            return;
        }
        
        if ( enemy.getWarriorID() == WarriorID.VAMPIRE || enemy.getWarriorID() == WarriorID.WEREWOLF || enemy.getWarriorID() == WarriorID.WOLF ) {
            enemy.setState(new SoldierState(enemy));
        }
        magician.getMagicPower().takeMana(manaCost);
    }
}

File: Army(c++)/Magic/Spell.cpp
#include "Spell.hpp"

Spell::Spell(int cost, int mana, int hp) : manaCost(cost), mana(mana), hitPoints(hp)  {}

Spell::~Spell() {}

void Spell::mAttack(Unit* enemy, SpellCaster* magician) {
    magician->getHealth().ensureIsAlive();
    magician->getMagicPower().ensureHaveIsMana();
    
    if ( magician->getMagicPower().getMana() < this->manaCost) {
        std::cout << "Not enough mana!" << std::endl;
        return;
    }
    
    if ( enemy->getWarriorID() == BERSERKER ) {
        std::cout << "Magic is powerless!" << std::endl;
        return;
    }
    
    if ( magician->getMagicianID() == SORCERER ) {
        enemy->getHealth().takeDamage(this->hitPoints);
    }
    
    if ( magician->getMagicianID() == HEALING ) {
        enemy->getHealth().takeDamage(this->hitPoints/2);
    }
    
    if ( magician->getMagicianID() == CHURCH_WORKER ) {
        if ( enemy->getWarriorID() == VAMPIRE || enemy->getWarriorID() == NECROMANCER ) {
            enemy->getHealth().takeDamage(this->hitPoints);
        } else {
            enemy->getHealth().takeDamage(this->hitPoints/2);
        }
    }
    
    magician->getMagicPower().takeMana(this->manaCost);
    
    enemy->notify();
}

void Spell::healHealth(Unit* other, SpellCaster* magician) {
    magician->getHealth().ensureIsAlive();
    
    if ( magician->getMagicPower().getMana() < this->manaCost) {
        std::cout << "Not enough mana!" << std::endl;
        return;
    }
    
    if ( other->getWarriorID() == BERSERKER || other->getWarriorID() == DEMON  ) {
        std::cout << "Magic is powerless!" << std::endl;
        return;
    }
    
    if ( magician->getMagicianID() == HEALING || magician->getMagicianID() == CHURCH_WORKER ) {
        other->getHealth().addHitPoints(this->hitPoints);
    } else {
        other->getHealth().addHitPoints(this->hitPoints/2);
    }
    
    magician->getMagicPower().takeMana(this->manaCost);
}

void Spell::healMagic(SpellCaster* other, SpellCaster* magician) {
    magician->getHealth().ensureIsAlive();
    other->getHealth().ensureIsAlive();
    
    if ( magician->getMagicPower().getMana() < this->manaCost) {
        std::cout << "Not enough mana!" << std::endl;
        return;
    }
    
    if ( magician->getMagicianID() == HEALING || magician->getMagicianID() == CHURCH_WORKER ) {
        other->getMagicPower().addMana(this->mana);
    } else {
        other->getMagicPower().addMana(this->mana/2);
    }
    
    magician->getMagicPower().takeMana(this->manaCost);
}

void Spell::backToHuman(Unit* enemy, SpellCaster* magician) {
    magician->getHealth().ensureIsAlive();
    magician->getMagicPower().ensureHaveIsMana();
    enemy->getHealth().ensureIsAlive();
    
    if ( magician->getMagicPower().getMana() < this->manaCost) {
        std::cout << "Not enough mana!" << std::endl;
        return;
    }
    
    if ( enemy->getWarriorID() == VAMPIRE || enemy->getWarriorID() == WEREWOLF || enemy->getWarriorID() == WOLF ) {
        enemy->transitionTo(new SoldierState(enemy));
    }
    
    magician->getMagicPower().takeMana(this->manaCost);
}
